### PR TITLE
Do not wrap errors in FailsafeException - Fallback won't handle Errors

### PR DESCRIPTION
**Pull request by gm2211**
https://github.com/jhalterman/failsafe/issues/182

### CODE REVIEW

File: src/main/java/net/jodah/failsafe/Fallback.java [hunk 1]
```diff
@@ -136,7 +136,10 @@ private Fallback(CheckedFunction<ExecutionAttemptedEvent, R> fallback, boolean a
   }
 
   public R apply(R result, Throwable failure, ExecutionContext context) throws Exception {
-    return fallback.apply(new ExecutionAttemptedEvent<R>(result, failure, context));
+    if (failure instanceof Error) {
```
Comment (gm2211) [inline src/main/java/net/jodah/failsafe/Fallback.java]: Fallback should not blindly handle errors - in fact I think it should bias toward throwing them.

However, we could in theory apply Failure policies here too to determine whether the Error should be handled or not (that'd probably require some refactoring).

Alternatively, we could accept a set of handled `Error` classes, this would create some duplication between the retry policy and the fallback, but would require fewer code changes.

I personally think this could be left as is, because if you really want to catch errors and have a fallback you can still wrap in a `try-catch` block.

**Comment by gm2211** [comment]
There are a couple of possible approaches to #182:

1. Don't catch Error - this is more robust, but would probably cause more debate (I really don't think Errors should be caught and handled, but people have different opinions, so they probably should have the option of handling errors if they are doing so intentionally)
2. [This PR] Keep the current framework, but do not wrap Error in `FailsafeException` and make Fallback always throw if the failure is an error

**Comment by gm2211** [comment]
@jhalterman Any chance you can take a look at this?

**Comment by jhalterman** [comment]
Sorry, I've been away recently.

However we handle this, or not, I want the approach to be consistent throughout Failsafe. Right now, you can configure any [FailurePolicy](http://jodah.net/failsafe/javadoc/net/jodah/failsafe/FailurePolicy.html) to handle any type of Throwable, including Fallback and RetryPolicy. So whatever we do should apply to all policies.

In theory, an `Error` is not recoverable, but in practice it's a bit more uncertain (ex: `IOError` and  `AssertionError`), so I'm happy to leave it to users to decide what to do when an `Error` is handled by their policy.

We may not want to wrap Errors at the top level with `FailsafeException`. But then again, this could lead to a somewhat inconsistent experience, where async execution results always come wrapped in a `CompletableFuture` that throws `ExecutionException` when you do a `get()`. Any thoughts on this?

Any thoughts on why not allowing Policies to handle Errors might be necessary?

**Comment by gm2211** [comment]
@jhalterman I've been extremely busy at work and haven't had time to come back and address your comment/fix this PR.

**TLDR: It's fine to let people handle Errors explicitly, but, unless handled, errors should not be wrapped in `FailsafeException` and `Fallback` should not swallow errors.**

I've given a little more thought about what spurred my initial reaction.

**Background:**
In the very distant past we have bitten by devs defaulting catch clauses to `Throwable` - e.g.
```
    catch(Throwable t) { 
        log.info("Error while doing X - returning default value", t);
        return defaultValue;
    }
```
which will catch `Error` too.
The dev most likely did not intend to catch `Error`s and handle them, or else they'd have a separate catch block for a specific error.
Even if the code is not swallowing the exception because it's being logged, JVM is being prevented from terminating, and devs would generally not notice, unless they are constantly monitoring logs.

We since switched to
```
    catch(RuntimeException e) { 
        log.info("Error while doing X - returning default value", e);
        metrics.recordFailure(...);
        return defaultValue;    
    }
```

The conclusion we reached is that while there are, rarely, circumstances in which it might make sense to catch `Error`, doing so as a "blanket" practice, can be very damaging.

**Issues**

_Issue 1_
Errors are always wrapped in `FailsafeException` even when not handled

image

fails with

image

So, all of a sudden, our `catch(RuntimeException e)` block gets implicitly converted back to `catch(Throwable t)` because we're using `Failsafe`.

_Potential Solutions for Issue 1_

1. Default to not handling errors. When a throwable is caught during when calling the provided checked supplier, check if it is among the handled exceptions and only then wrap it in `FailsafeException`
2. Default to not handling errors. Always wrap in `FailsafeException`, but as the last step check before throwing: if no handlers for contained type and if `Error`, throw the raw `Error`.

_Issue 2_

Fallback will swallow exceptions as well as errors.

image

Fails with 

image

Again, this brings us back to the blanket catching errors behavior that is pretty dangerous.

_Potential Solutions for Issue 2_

1. Change the behavior of `Fallback` to always throw on error
2. Same as 1. but add `Fallback.handling(WhateverError.class).of(() -> value)`
3. Combine `Fallback` with `RetryPolicy` - e.g. `RetryPolicy.handle(...).fallback(...)`, this way the fallback is tied to explicitly handled situations.
 
**Bottom line**: I think it's ok to allow users of this library to handle errors if they choose to do so (though it's a tricky territory), but I think it would be better if no construct swallowed `Error` without explicit and obvious user action (i.e. I explicitly write `.handle(SomeError.class)`).

**Comment by jhalterman** [comment]
This was fixed in 2.3.0 - `Error`s are no longer wrapped, but users can still handle them in their policies the same as `Exceptions` if they wish.
